### DISCUSSION TITLE
generic constrain to allow only ref structs

### DISCUSSION THREAD

**Discussion by BarionLP** [dotnet/csharplang#8601] [General]
Not sure if I'm missing something but with the new `allow ref struct` thing it should also be possible to create a constrain that only allows ref structs.
e.g. I was just trying to add this one of my projects and came across multiple situations where I need different behavior for ref structs compared to normal structs and classes. but right now i can only differenciate between struct and class.
Is this a design oversight? Are there plans to add this in C# 14? Or was this specifically rejected?

Edit for clarity:
Currently this is valid C#:
```csharp
"".Foo(); // ClassWrapperExtensions.ToWrapper
1.Foo(); // StructWrapperExtensions.ToWrapper

public static class ClassWrapperExtensions
{
    public static ClassWrapper<T> ToWrapper<T>(this T value) where T : class { }
}

public static class StructWrapperExtensions
{
    public static StructWrapper<T> ToWrapper<T>(this T value) where T : struct { }
}
```
I want to extend this to also cover ref structs (in my use case i only need ref struct or not ref struct)
```csharp
"".Foo(); // ClassWrapperExtensions.ToWrapper
1.Foo(); // StructWrapperExtensions.ToWrapper
Span<byte>.Empty.Foo(); // RefWrapperExtensions.ToWrapper

public static class ClassWrapperExtensions
{
    public static ClassWrapper<T> ToWrapper<T>(this T value) where T : class { }
}

public static class StructWrapperExtensions
{
    public static StructWrapper<T> ToWrapper<T>(this T value) where T : struct { }
}

public static class RefWrapperExtensions
{
    public static RefWrapper<T> ToWrapper<T>(this T value) where T : ref struct { }
}
```

**Comment by CyrusNajmabadi** [comment]
What's your use case for this?

**Comment by BarionLP** [comment]
> What's your use case for this?

specifically i have two wrapper types, one feature rich for classes/structs and a stripped down version for ref structs. now i need a generic extension method that wraps ref structs in the stripped version and the others in the full version without any ambiguity. Of course I can just name it differently but then I have the problem that users might accidentally create a stripped version with a type that supports full functionality. Both versions are fully stack allocated (there is no notable overhead of the full version). So there is no reason to use the stripped version for classes/structs it's just that some features are not available, like obtaining the default equality comparer.

**Reply by CyrusNajmabadi** [reply]
In this case, can't you just have a static constructor in the type that fails if you specify a type arg that you don't want to support?

**Reply by BarionLP** [reply]
still wouldn't solve the ambiguity problem and it doesn't seem very pretty especially since i don't get compiler errors
but it might be a workaround for now

**Reply by CyrusNajmabadi** [reply]
> still wouldn't solve the ambiguity problem

I don't know what this is referring to.  Or why a 'requires ref struct' sort of constraint would solve things for now.

> and it doesn't seem very pretty especially since i don't get compiler errors

You can always add compiler errors through an analyzer.  That is our *recommended* path for "the language allows this, but in my domain i want to disallow it in certain circumstances".

**Reply by jaredpar** [reply]
Can you share code that demonstrates _what_ you want to write? 

> now i need a generic extension method that wraps ref structs in the stripped version and the others in the full version without any ambiguity.

When I read this I feel like you're asking for a way to have overloads that behave differently for `ref struct` and everything else. That's not possible. But what you could write is something like 

```csharp
void M<T>(T value) where T : allows ref struct {
  if (typeof(T).IsByRef) { 
    // ref struct path
  }
  else { 
   // other path
  }
}
```

**Reply by BarionLP** [reply]
> Can you share code that demonstrates _what_ you want to write?

i updated the description, i hope that makes it clearer

**Reply by BarionLP** [reply]
> ```cs
> if (typeof(T).IsByRef) { 
>     // ref struct path
>   }
>   else { 
>    // other path
>   }
> ```

i assume the compiler doesn't let me box the value in the else branch, also i need different return types

**Reply by jaredpar** [reply]
Correct that won't let you box because checking the value doesn't change the understood constraints. Think what you're looking for here is bridge constraints proposal 

https://github.com/dotnet/csharplang/discussions/6308

This is what would let us basically change the constraints on a `T` within a method as the actual constraints are evaluated.

**Reply by BarionLP** [reply]
> Think what you're looking for here is bridge constraints proposal

still need a different return type

**Reply by jaredpar** [reply]
I'm unsure though how you expect to use this. As I mentioned, you cannot overload methods based on constraints. So even if this constraint existed what would you do? Essentially, pretend for a moment that we had `where T: ref struct` and it was only satisfied for actual `ref struct`. What is the API that you would like to write?

**Reply by BarionLP** [reply]
> What is the API that you would like to write?

doesn't the updated description make it clear?
its a `ToOption` extension method that either creates a `struct Option` (can be used in async / yield return) or a `ref struct RefOption` (has limitations). 

As shown in the description it is possible to have a `ClassOption` and `StructOption` and have the compiler pick the correct extension method but i only need the ref struct and not ref struct separation

**Reply by colejohnson66** [reply]
Understandable, but is there a reason you need a `RefOption<T>`? Why wouldn't `Option<T> where T : allows ref struct` suffice?

**Reply by BarionLP** [reply]
> Understandable, but is there a reason you need a `RefOption<T>`? Why wouldn't `Option<T> where T : allows ref struct` suffice?

so it can be used in async / yield return.
Also i need the `EqualityComparer<T>.Default` and override `GetHashCode`/`ToString`

**Comment by TahirAhmadov** [comment]
I think what you may be looking for is generic specialization:
```cs
class Wrapper<T> where T : allows ref struct
{
  // stripped down 
  public void ForEverything() { ... }
  // full featured
  public void ForClass() where T : class { ... }
  public void ForStruct() where T : struct { ... }
}
```
There probably are discussions (or even issues) requesting this.

**Reply by BarionLP** [reply]
> I think what you may be looking for is generic specialization:

yes, that would solve my problem, but more like this
```csharp
class Wrapper<T> where T : allows ref struct
{
  public void ForEverything() { ... }
  public void ForClass() where T : not ref struct { ... }
  public void ForRefStruct() where T : only ref struct { ... }
}
```
my current scenario looks like this:
```csharp
public record struct Wrapper<T>(T Val);
public ref struct RefWrapper<T>(T Val) where T : allows ref struct;
```
but i want it to look like this
```csharp
public struct Wrapper<T>(T Val); //this can implement stuff that does not work with ref structs and used across async / yield return
public ref struct RefWrapper<T>(T Val) where T : ref struct; //only ref structs
```
with this being valid
```csharp
Create(1); //struct to Wrapper
Create(""); //class to Wrapper
Create(Span<byte>.Empty); // ref struct to RefWrapper

public static Wrapper Create(T val) => new Wrapper<T>(val);
public static RefWrapper Create(T val) where T : ref struct => new RefWrapper<T>(val);
```
even without generic specialization it should be possible to have a `CreateRef` method that only accepts ref structs

**Reply by BarionLP** [reply]
I was confused because this works
```csharp
public static Wrapper<T> Create<T>(T? val) where T : class=> new Wrapper<T>(val);
public static Wrapper<T> Create<T>(T? val) where T : struct => new Wrapper<T>(val.Value);
```
but `T?` becomes `Nullable<T>` which is a different signature

**Reply by BarionLP** [reply]
functionality wise this isn't necessary right now (just with generic specialization this becomes mandatory)
but it would be nice to have to prevent unintended usage

**Reply by KennethHoff** [reply]
This sounds like you just want peace of mind as it seemingly wouldn't affect the API in any way, which is what analyzers were designed for. Create a `[OnlyRefStructAttribute]` which can be applied to type parameters and a corresponding analyzer that verifies it

**Reply by BarionLP** [reply]
> This sounds like you just want peace of mind as it seemingly wouldn't affect the API in any way, which is what analyzers were designed for. Create a `[OnlyRefStructAttribute]` which can be applied to type parameters and a corresponding analyzer that verifies it

almost, it would affect extension method overload resolution (can i have one for struct and one for class in different classes but not for ref struct)
https://github.com/dotnet/csharplang/discussions/8601#discussioncomment-11303234

**Reply by CyrusNajmabadi** [reply]
> it would affect extension method overload resolution

Constraints don't affect overload resolution.

**Reply by BarionLP** [reply]
> > it would affect extension method overload resolution
> 
> Constraints don't affect overload resolution.

it does with extensions methods
```csharp
"".Foo(); //A.Foo
1.Foo(); //B.Foo

public static class A
{
    public static void Foo<T>(this T value) where T : class { }
}

public static class B
{
    public static void Foo<T>(this T value) where T : struct { }
}
```
i now want the same for ref struct

**Comment by huoyaoyuan** [comment]
Constraint that **restricting scopes only** are mostly suitable for analyzers. Constraints built-in into language should enable **additional operations under the constraint**.

Is there any operation you can do with an instance of `ref struct`, but not with a regular `struct`?

**Reply by BarionLP** [reply]
> Is there any operation you can do with an instance of `ref struct`, but not with a regular `struct`?

no, I don't think so. 
The only usecase would be overload resolution
```csharp
1.Foo(); //this is ambiguous right now
Span<byte>.Empty.Foo();

public static class A
{
    public static void Foo<T>(this T value) 
    { 
        // something that cannot be done with ref structs and should be done for everything but ref structs
    }
}

public static class B
{
    public static void Foo<T>(this T value)  where T : allows ref struct //this would have to be ref struct only
    {
        //fallback logic for ref structs
     }
}
```
This would work with one for `: struct` and one for `: class`, which is why i wanted to do it this way.
`B.Foo` can of course just be `A.FooRef` and everything would work fine, it just looks ugly

**Comment by ufcpp** [comment]
Could better "[Bestest Betterness](https://github.com/dotnet/csharplang/blob/25c33b53f24f4d365412d744abe42e865a11776b/meetings/2018/LDM-2018-02-26.md#bestest-betterness)" solve this issue?

```cs
using System.Runtime.CompilerServices;

"".M();

// Ambiguous before C# 7.3, but now resolvable. (Bestest betterness)
1.M();

// Improving "Bestest Betterness" rule could solve this issue?
((Span<int>)[]).M();

static class Extensions
{
    public struct ClassDummy;
    public struct StructDummy;
    public struct RefStructDummy;

    public static void M<T>(this T x, ClassDummy _ = default) where T : class { }
    public static void M<T>(this T x, StructDummy _ = default) where T : struct { }

    [OverloadResolutionPriority(-1)]
    public static void M<T>(this T x, RefStructDummy _ = default) where T : struct, allows ref struct { }
}
```

**Reply by BarionLP** [reply]
that looks interesting, it might do, (despite being ugly)
